### PR TITLE
Fix error of __list_find

### PR DESCRIPTION
**Pull request by RinHizakura**
The node on the linked list should follow the ordering by its key. However, the function `list_insert` doesn't actually do this because of the wrong implementation of `__list_find`.

There are three main fixes in this patch. First, once we should do synchronization, we have to iterate the list again but not return from the function. Second, the while loop should iterate until the last node but not the second last one. Finally,
under the normal situation that the tail node with key `UINTPTR_MAX` should not be deleted (unless the end of the program), some codes are never executed which can be removed.

### CODE REVIEW

File: hp_list/main.c [hunk 1]
```diff
@@ -230,14 +230,10 @@ static bool __list_find(list_t *list,
...
     while (true) {
-        if (!get_unmarked_node(curr))
-            return false;
         next = (list_node_t *) atomic_load(&get_unmarked_node(curr)->next);
         (void) list_hp_protect_ptr(list->hp, HP_NEXT, get_unmarked(next));
         if (atomic_load(&get_unmarked_node(curr)->next) != (uintptr_t) next)
-            break;
-        if (get_unmarked(next) == atomic_load((atomic_uintptr_t *) &list->tail))
-            break;
+            goto try_again;
```
Comment (jserv) [inline hp_list/main.c:240]: You can add some comments around this part.

File: hp_list/main.c [hunk 2]
```diff
@@ -230,14 +230,13 @@ static bool __list_find(list_t *list,
     if (atomic_load(prev) != get_unmarked(curr))
         goto try_again;
     while (true) {
-        if (!get_unmarked_node(curr))
-            return false;
         next = (list_node_t *) atomic_load(&get_unmarked_node(curr)->next);
         (void) list_hp_protect_ptr(list->hp, HP_NEXT, get_unmarked(next));
+        /* If the relationship between prev / curr and curr / next is different
```
Comment (RinHizakura) [inline hp_list/main.c]: I have no confidence that the comment is precise enough to describe why we should `goto try_agin`.  Any advice if I can improve the comment?

**Comment by jserv** [comment]
I would expect the following comments:
> On a CAS failure, the search function, "__list_find",  will simply have to
> go backwards in the list until an unmarked element is found from which 
> he search in increasing key order can be started.

> For the implementation to be correct, the predecessor pointers only
> have to fulfill that any item always has a path back to the head of the
list. When a new item is inserted into the list, the predecessor pointer
of the successor is set with an atomic store to point to the new element.
> Likewise, when an item is removed, the predecessor pointer of the
> successor item is updated to skip the removed item. This suffices to
> maintain the invariant, but through long sequences of concurrent
> insertions and deletions, backwards pointers can become imprecise
> in that they skip many items that are actually in the list.

My implementation was a functional rewriting of paper [A more Pragmatic Implementation of the Lock-free, Ordered, Linked List](https://arxiv.org/abs/2010.15755). You can check its implementations and benchmark programs: [lockfree-linked-list](https://github.com/parlab-tuwien/lockfree-linked-list).

**Comment by RinHizakura** [comment]
Thanks for the information. I changed the comment according to yours.

File: hp_list/main.c [hunk 3]
```diff
@@ -230,14 +230,13 @@ static bool __list_find(list_t *list,
     if (atomic_load(prev) != get_unmarked(curr))
         goto try_again;
     while (true) {
-        if (!get_unmarked_node(curr))
-            return false;
         next = (list_node_t *) atomic_load(&get_unmarked_node(curr)->next);
         (void) list_hp_protect_ptr(list->hp, HP_NEXT, get_unmarked(next));
+        /* On a CAS failure, the search function, "__list_find", will simply have to
```
Comment (jserv) [inline hp_list/main.c]: Use the consistent style on comments. That is,
```c
/* On a CAS failure, ...
 * ...
 */
```

Comment (jserv) [reply hp_list/main.c]: should be
> the search function, "__list_find," will ...

See https://www.grammarly.com/blog/quotation-marks/

**Pull request merged**